samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java

```diff
@@ -255,6 +261,31 @@ private void blockIfBusy(IncomingMessageEnvelope envelope) {
     }
   }
 
+  /**
+   * When there are no new messages to process and the run loop has not been resumed since the last time this code was
+   * run, block the AsyncRunLoop thread for a short while. This will prevent this task from spinning when it has no work
+   * to do. If a task worker finishes or window/commit completes before the timeout it will resume the AsyncRunLoop
+   * thread immediately. This performs a similar function to what the poll timeout in the SystemConsumers.choose method
+   * provides for the single threaded RunLoop code, but, in the async case we need the ability to resume any time a task
+   * worker completes processing a message. In the Async case, that event will allow the task worker to start processing
+   * a message that may have already been chosen.
+   */
+  private void blockIfNoWork(IncomingMessageEnvelope envelope) {
+    synchronized (latch) {
+      if ((envelope != null) || runLoopResumedSinceLastChecked) {
```

Comment:
I think that @xinyuiscool mentioned it. I also just realized that we can merge the two block functions into one, if we do not wait on the consumer.poll:
1) make sure the exit/return condition in the synchronized block is: (there is a pending messages in the queue or there is an new envelope returned from poll) && (the corresponding task worker is available: i.e. not busy);
2) otherwise, wait on latch with task.max.idle.ms. 

Does that sound good?

Reply:
I am also not sure how to implement the exact exit/return logic you suggest above.  It sounds like that would (at least) require checking all the pendingEnvelopeQueues for messages and matching them to the idle workers.  It seems like I might also need to check for pending windows and/or commits.  It seems easier/safer to just call run() on all the task workers if at least one task worker is idle (the current logic).

I am also not sure what I should do if the "combined trigger" wait you suggest times out?  The current blockIfBusy can continue to wait for an unlimited amount of time and can immediately wait again if after the notify all the tasks are still busy.  Are you suggesting changing this logic to block at most once for a fixed amount of time (the blockIfNoWork condition should block at most once)?

What I can do is check the two existing conditions used in blockIfNoWork and blockIfBusy serially inside a common latch (most likely called between the choose and run steps).  I might be able to combine them into a single "if" check it the blockIfBusy logic is relaxed to block at most once and for a fixed amount of time.  That might cause the run loop to waste cycles in a very busy application.

Comments before I proceed?

Reply:
I did not make "runLoopResumedSinceLastChecked" volatile because I only set and check that variable inside code blocks synchronized on "latch", but, it could not hurt to make it volatile (and if my assumption is wrong it is required).  My assumption was that the synchronized/wait/notify semantics would ensure each thread saw the latest value.  Change is on the way.

Reply:
@jwlent55 : it will be super nice if you can help move your logic in blockInNoWork into blockIfBusy which do the things you mentioned: check the two existing conditions used in blockIfNoWork and blockIfBusy serially inside a common latch. That will make the code easier to follow and reason since the latch happens in one place. The other part looks good to me. Thanks.

Reply:
Done.